### CapStone Project Week 5
## Introduction

Moving to a new city is one of the most stressfull events in someone's life. As you leave your home, you don't know what you'll find on the other side. Where will you shop your grosseries, which school your kids will go to, will you find new friends? How long will it take before you feel at home again?

In this project we'll suggest where someone living in Paris should move in Brussels.



## Data

We'll use the postal codes of both cities to identify their neighborhoods. 
The list of postal codes of Paris may be found here: https://www.laposte.fr/particulier/outils/trouver-un-code-postal
Those of Brussels can be found here: https://www.bpost.be/site/fr/envoyer/adressage/rechercher-un-code-postal

Let's import them and see what they look like:

In [1]:
import numpy as np

In [2]:
# The code was removed by Watson Studio for sharing.

Code Postal  Ville
0         75001  PARIS
1         75002  PARIS
2         75003  PARIS
3         75004  PARIS
4         75005  PARIS
5         75006  PARIS
6         75007  PARIS
7         75008  PARIS
8         75009  PARIS
9         75010  PARIS
10        75011  PARIS
11        75012  PARIS
12        75013  PARIS
13        75014  PARIS
14        75015  PARIS
15        75016  PARIS
16        75017  PARIS
17        75018  PARIS
18        75019  PARIS
19        75020  PARIS
20        75116  PARIS

Postal Code ***75116*** actually corresponds to the same neighborhoods as ***75016*** so we'll drop it

In [3]:
paris_codes_df.drop([20], inplace = True)
paris_codes_df

Code Postal  Ville
0         75001  PARIS
1         75002  PARIS
2         75003  PARIS
3         75004  PARIS
4         75005  PARIS
5         75006  PARIS
6         75007  PARIS
7         75008  PARIS
8         75009  PARIS
9         75010  PARIS
10        75011  PARIS
11        75012  PARIS
12        75013  PARIS
13        75014  PARIS
14        75015  PARIS
15        75016  PARIS
16        75017  PARIS
17        75018  PARIS
18        75019  PARIS
19        75020  PARIS

In [4]:
paris_codes_df.rename(columns={"Code Postal": "Postal Code"}, inplace = True)
paris_codes_df.head()

Postal Code  Ville
0        75001  PARIS
1        75002  PARIS
2        75003  PARIS
3        75004  PARIS
4        75005  PARIS

The neighborhoods in Paris are known by their numbers. I'll add that information to the data and drop the column 'Ville' because it doesn't give much information.

In [5]:
paris_codes_df['Neighborhood'] = ["Paris 1", "Paris 2", "Paris 3", "Paris 4", "Paris 5",
                                 "Paris 6", "Paris 7", "Paris 8", "Paris 9", "Paris 10",
                                 "Paris 11", "Paris 12", "Paris 13", "Paris 14", "Paris 15",
                                 "Paris 16", "Paris 17", "Paris 18", "Paris 19", "Paris 20"]
paris_codes_df.drop("Ville", axis = 1, inplace=True)
paris_codes_df

Postal Code Neighborhood
0         75001      Paris 1
1         75002      Paris 2
2         75003      Paris 3
3         75004      Paris 4
4         75005      Paris 5
5         75006      Paris 6
6         75007      Paris 7
7         75008      Paris 8
8         75009      Paris 9
9         75010     Paris 10
10        75011     Paris 11
11        75012     Paris 12
12        75013     Paris 13
13        75014     Paris 14
14        75015     Paris 15
15        75016     Paris 16
16        75017     Paris 17
17        75018     Paris 18
18        75019     Paris 19
19        75020     Paris 20

In [6]:
print('There are {} neighborhoods in Paris'.format(paris_codes_df.shape[0]))

There are 20 neighborhoods in Paris


Let's see the postal codes information for the city of Brussels

In [7]:

body = client_75203cb60a4e41d4bb1debcd2578d37c.get_object(Bucket='machinelearning-donotdelete-pr-7gneoytsamib8b',Key='brussels_postal_codes.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

brussels_codes_df = pd.read_csv(body,';')
brussels_codes_df


Code postal               Localité Sous-commune     Commune principale  \
0          1000              Bruxelles          Non              BRUXELLES   
1          1020                 Laeken          Oui              BRUXELLES   
2          1030             Schaerbeek          Non             SCHAERBEEK   
3          1040              Etterbeek          Non              ETTERBEEK   
4          1050                Ixelles          Non                IXELLES   
5          1060           Saint-Gilles          Non           SAINT-GILLES   
6          1070             Anderlecht          Non             ANDERLECHT   
7          1080   Molenbeek-Saint-Jean          Non   MOLENBEEK-SAINT-JEAN   
8          1081             Koekelberg          Non             KOEKELBERG   
9          1082  Berchem-Sainte-Agathe          Non  BERCHEM-SAINTE-AGATHE   
10         1083              Ganshoren          Non              GANSHOREN   
11         1090                  Jette          Non                  JETTE   
12         1120    Neder-Over-Heembeek          Oui              BRUXELLES   
13         1130                  Haren          Oui              BRUXELLES   
14         1140                  Evere          Non                  EVERE   
15         1150    Woluwe-Saint-Pierre          Non    WOLUWE-SAINT-PIERRE   
16         1160              Auderghem          Non              AUDERGHEM   
17         1170    Watermael-Boitsfort          Non    WATERMAEL-BOITSFORT   
18         1180                  Uccle          Non                  UCCLE   
19         1190                 Forest          Non                 FOREST   
20         1200   Woluwe-Saint-Lambert          Non   WOLUWE-SAINT-LAMBERT   
21         1210  Saint-Josse-Ten-Noode          Non  SAINT-JOSSE-TEN-NOODE   

   Rôle linguistique  
0            NF - FN  
1            NF - FN  
2            NF - FN  
3            NF - FN  
4            NF - FN  
5            NF - FN  
6            NF - FN  
7            NF - FN  
8            NF - FN  
9            NF - FN  
10           NF - FN  
11           NF - FN  
12           NF - FN  
13           NF - FN  
14           NF - FN  
15           NF - FN  
16           NF - FN  
17           NF - FN  
18           NF - FN  
19           NF - FN  
20           NF - FN  
21           NF - FN

We'll keep the first two columns of the data frame

In [8]:
brussels_codes_df.drop(columns=['Sous-commune','Commune principale','Rôle linguistique'], inplace=True)
brussels_codes_df.head()

Code postal    Localité
0         1000   Bruxelles
1         1020      Laeken
2         1030  Schaerbeek
3         1040   Etterbeek
4         1050     Ixelles

I'll rename the columns to 'Postal Code' and 'Neighborhood'

In [9]:
brussels_codes_df.rename(columns={"Code postal": "Postal Code", "Localité":"Neighborhood"}, inplace=True)
brussels_codes_df.head()

Postal Code Neighborhood
0         1000    Bruxelles
1         1020       Laeken
2         1030   Schaerbeek
3         1040    Etterbeek
4         1050      Ixelles

In [10]:
print('There are {} neighborhoods in Brussels'.format(brussels_codes_df.shape[0]))

There are 22 neighborhoods in Brussels


We need the coordinates of each neighborhood in order to use the Foursquare API

### Paris neighborhoods coordinates

In [11]:

body = client_75203cb60a4e41d4bb1debcd2578d37c.get_object(Bucket='machinelearning-donotdelete-pr-7gneoytsamib8b',Key='paris_coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

paris_coordinates_df = pd.read_csv(body,';')
paris_coordinates_df


Postal Code   Latitude  Longitude
0         75001  48.860074   2.341092
1         75002  48.866729   2.340470
2         75003  48.863873   2.361786
3         75004  48.856053   2.356322
4         75005  48.846198   2.344591
5         75006  48.850714   2.332658
6         75007  48.856860   2.320032
7         75008  48.877364   2.317352
8         75009  48.872598   2.340453
9         75010  48.871792   2.357372
10        75011  48.857713   2.380543
11        75012  48.841114   2.388338
12        75013  48.832314   2.355485
13        75014  48.832866   2.326794
14        75015  48.841753   2.299553
15        75016  48.863864   2.276589
16        75017  48.884331   2.322106
17        75018  48.892618   2.345112
18        75019  48.882586   2.382156
19        75020  48.865088   2.398986

### Brussels neighborhoods coordinates

In [12]:

body = client_75203cb60a4e41d4bb1debcd2578d37c.get_object(Bucket='machinelearning-donotdelete-pr-7gneoytsamib8b',Key='brussels_coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

brussels_coordinates_df = pd.read_csv(body,';')
brussels_coordinates_df


Postal Code   Latitude  Longitude
0          1000  50.846714   4.352514
1          1020  50.876340   4.347956
2          1030  50.868056   4.373598
3          1040  50.834770   4.387934
4          1050  50.832551   4.367482
5          1060  50.824793   4.346348
6          1070  50.839148   4.328931
7          1080  50.854850   4.340254
8          1081  50.860935   4.331102
9          1082  50.864812   4.294890
10         1083  50.869365   4.308765
11         1090  50.875802   4.324237
12         1120  50.897867   4.385816
13         1130  50.892091   4.418116
14         1140  50.872615   4.402962
15         1150  50.837378   4.427669
16         1160  50.817025   4.427771
17         1170  50.798056   4.417055
18         1180  50.803547   4.333215
19         1190  50.811284   4.319380
20         1200  50.847449   4.435074
21         1210  50.850773   4.369453

For each city I'll merge the data frames containing neighborhoods names and coordinates.

In [13]:
paris_df = paris_codes_df.merge(paris_coordinates_df, left_on='Postal Code', right_on='Postal Code')
paris_df

Postal Code Neighborhood   Latitude  Longitude
0         75001      Paris 1  48.860074   2.341092
1         75002      Paris 2  48.866729   2.340470
2         75003      Paris 3  48.863873   2.361786
3         75004      Paris 4  48.856053   2.356322
4         75005      Paris 5  48.846198   2.344591
5         75006      Paris 6  48.850714   2.332658
6         75007      Paris 7  48.856860   2.320032
7         75008      Paris 8  48.877364   2.317352
8         75009      Paris 9  48.872598   2.340453
9         75010     Paris 10  48.871792   2.357372
10        75011     Paris 11  48.857713   2.380543
11        75012     Paris 12  48.841114   2.388338
12        75013     Paris 13  48.832314   2.355485
13        75014     Paris 14  48.832866   2.326794
14        75015     Paris 15  48.841753   2.299553
15        75016     Paris 16  48.863864   2.276589
16        75017     Paris 17  48.884331   2.322106
17        75018     Paris 18  48.892618   2.345112
18        75019     Paris 19  48.882586   2.382156
19        75020     Paris 20  48.865088   2.398986

In [14]:
brussels_df = brussels_codes_df.merge(brussels_coordinates_df, left_on='Postal Code', right_on='Postal Code')
brussels_df

Postal Code           Neighborhood   Latitude  Longitude
0          1000              Bruxelles  50.846714   4.352514
1          1020                 Laeken  50.876340   4.347956
2          1030             Schaerbeek  50.868056   4.373598
3          1040              Etterbeek  50.834770   4.387934
4          1050                Ixelles  50.832551   4.367482
5          1060           Saint-Gilles  50.824793   4.346348
6          1070             Anderlecht  50.839148   4.328931
7          1080   Molenbeek-Saint-Jean  50.854850   4.340254
8          1081             Koekelberg  50.860935   4.331102
9          1082  Berchem-Sainte-Agathe  50.864812   4.294890
10         1083              Ganshoren  50.869365   4.308765
11         1090                  Jette  50.875802   4.324237
12         1120    Neder-Over-Heembeek  50.897867   4.385816
13         1130                  Haren  50.892091   4.418116
14         1140                  Evere  50.872615   4.402962
15         1150    Woluwe-Saint-Pierre  50.837378   4.427669
16         1160              Auderghem  50.817025   4.427771
17         1170    Watermael-Boitsfort  50.798056   4.417055
18         1180                  Uccle  50.803547   4.333215
19         1190                 Forest  50.811284   4.319380
20         1200   Woluwe-Saint-Lambert  50.847449   4.435074
21         1210  Saint-Josse-Ten-Noode  50.850773   4.369453

### Foursquare API

We'll use Foursquare API in order to characterise each neighborhood of both cities

In [15]:
!pip -q install folium
import folium # map rendering library
print('folium imported')

folium imported


In [16]:
!pip install geopy
from geopy.geocoders import Nominatim
print('Nominatum imported')

Nominatum imported


In [17]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [18]:
address_p = 'Paris, FR'

geolocator_p = Nominatim(user_agent="paris_explorer")
location_p = geolocator_p.geocode(address_p)
latitude_p = location_p.latitude
longitude_p = location_p.longitude
print('The geograpical coordinate of Paris are {}, {}.'.format(latitude_p, longitude_p))

The geograpical coordinate of Paris are 48.8566969, 2.3514616.


In [19]:
address_b = 'Brussels, BE'

geolocator_b = Nominatim(user_agent="brussels_explorer")
location_b = geolocator_b.geocode(address_b)
latitude_b = location_b.latitude
longitude_b = location_b.longitude
print('The geograpical coordinate of Brussels are {}, {}.'.format(latitude_b, longitude_b))

The geograpical coordinate of Brussels are 50.8465573, 4.351697.


In [20]:
# create map of Paris using latitude and longitude values
map_of_paris = folium.Map(location=[latitude_p, longitude_p], zoom_start=12)

# add markers to map
for lat, lng, neighborhood in zip(paris_df['Latitude'], paris_df['Longitude'], paris_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_of_paris)  
    
map_of_paris

In [21]:
# create map of Brussels using latitude and longitude values
map_of_brussels = folium.Map(location=[latitude_b, longitude_b], zoom_start=12)

# add markers to map
for lat, lng, neighborhood in zip(brussels_df['Latitude'], brussels_df['Longitude'], brussels_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_of_brussels)  
    
map_of_brussels

I'll join the neighborhoods of both cities in one data frame:

In [50]:
frames = [paris_df, brussels_df]
both_cities_neigh_df = pd.concat(frames)
both_cities_neigh_df.tail()

Postal Code           Neighborhood   Latitude  Longitude
17         1170    Watermael-Boitsfort  50.798056   4.417055
18         1180                  Uccle  50.803547   4.333215
19         1190                 Forest  50.811284   4.319380
20         1200   Woluwe-Saint-Lambert  50.847449   4.435074
21         1210  Saint-Josse-Ten-Noode  50.850773   4.369453

I'll now use Foursquare to get the venues of each neighborhood for both cities.

In [22]:
# The code was removed by Watson Studio for sharing.

In [23]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

In [24]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [25]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        if (len(requests.get(url).json()["response"])== 0):
            results = {}
        else:
            results = requests.get(url).json()["response"]['groups'][0]['items']
        #results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [26]:
paris_venues = getNearbyVenues(names=paris_df['Neighborhood'],
                                  latitudes=paris_df['Latitude'],
                                  longitudes=paris_df['Longitude'])

Paris 1
Paris 2
Paris 3
Paris 4
Paris 5
Paris 6
Paris 7
Paris 8
Paris 9
Paris 10
Paris 11
Paris 12
Paris 13
Paris 14
Paris 15
Paris 16
Paris 17
Paris 18
Paris 19
Paris 20


In [27]:
print(paris_venues.shape)
paris_venues.head()

(1357, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0      Paris 1              48.860074                2.341092   
1      Paris 1              48.860074                2.341092   
2      Paris 1              48.860074                2.341092   
3      Paris 1              48.860074                2.341092   
4      Paris 1              48.860074                2.341092   

                                               Venue  Venue Latitude  \
0                              Cour Carrée du Louvre       48.860360   
1                                    Place du Louvre       48.859841   
2  Église Saint-Germain-l'Auxerrois (Église Saint...       48.859520   
3                                      Coffee Crêpes       48.858841   
4                                  Boutique yam'Tcha       48.861710   

   Venue Longitude      Venue Category  
0         2.338543    Pedestrian Plaza  
1         2.340822               Plaza  
2         2.341306              Church  
3         2.340802         Coffee Shop  
4         2.342380  Chinese Restaurant

In [28]:
brussels_venues = getNearbyVenues(names=brussels_df['Neighborhood'],
                                  latitudes=brussels_df['Latitude'],
                                  longitudes=brussels_df['Longitude'])

Bruxelles
Laeken
Schaerbeek
Etterbeek
Ixelles
Saint-Gilles
Anderlecht
Molenbeek-Saint-Jean
Koekelberg
Berchem-Sainte-Agathe
Ganshoren
Jette
Neder-Over-Heembeek
Haren
Evere
Woluwe-Saint-Pierre
Auderghem
Watermael-Boitsfort
Uccle
Forest
Woluwe-Saint-Lambert
Saint-Josse-Ten-Noode


In [29]:
print(brussels_venues.shape)
brussels_venues.head()

(733, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Bruxelles              50.846714                4.352514   
1    Bruxelles              50.846714                4.352514   
2    Bruxelles              50.846714                4.352514   
3    Bruxelles              50.846714                4.352514   
4    Bruxelles              50.846714                4.352514   

                                               Venue  Venue Latitude  \
0            Grand Place / Grote Markt (Grote Markt)       50.846776   
1                               AB Ancienne Belgique       50.847143   
2  Galerie de la Reine / Koninginnegalerij (Konin...       50.847297   
3                            Aux Merveilleux de Fred       50.848117   
4                                  Mary Chocolaterie       50.847541   

   Venue Longitude  Venue Category  
0         4.352481           Plaza  
1         4.349485    Concert Hall  
2         4.354601   Shopping Mall  
3         4.352141    Dessert Shop  
4         4.354785  Chocolate Shop

In [30]:
paris_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                         
Paris 1                          60                      60     60   
Paris 10                         57                      57     57   
Paris 11                         84                      84     84   
Paris 12                         42                      42     42   
Paris 13                         53                      53     53   
Paris 14                         45                      45     45   
Paris 15                         44                      44     44   
Paris 16                         34                      34     34   
Paris 17                         99                      99     99   
Paris 18                         58                      58     58   
Paris 19                         32                      32     32   
Paris 2                         100                     100    100   
Paris 20                         60                      60     60   
Paris 3                          95                      95     95   
Paris 4                         100                     100    100   
Paris 5                         100                     100    100   
Paris 6                          61                      61     61   
Paris 7                          78                      78     78   
Paris 8                         100                     100    100   
Paris 9                          55                      55     55   

              Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                   
Paris 1                   60               60              60  
Paris 10                  57               57              57  
Paris 11                  84               84              84  
Paris 12                  42               42              42  
Paris 13                  53               53              53  
Paris 14                  45               45              45  
Paris 15                  44               44              44  
Paris 16                  34               34              34  
Paris 17                  99               99              99  
Paris 18                  58               58              58  
Paris 19                  32               32              32  
Paris 2                  100              100             100  
Paris 20                  60               60              60  
Paris 3                   95               95              95  
Paris 4                  100              100             100  
Paris 5                  100              100             100  
Paris 6                   61               61              61  
Paris 7                   78               78              78  
Paris 8                  100              100             100  
Paris 9                   55               55              55

In [31]:
print('There are {} uniques categories in Paris.'.format(len(paris_venues['Venue Category'].unique())))

There are 199 uniques categories in Paris.


In [32]:
brussels_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                  
Anderlecht                                23                      23     23   
Auderghem                                 43                      43     43   
Berchem-Sainte-Agathe                     14                      14     14   
Bruxelles                                100                     100    100   
Etterbeek                                 42                      42     42   
Evere                                     16                      16     16   
Forest                                    15                      15     15   
Ganshoren                                 30                      30     30   
Haren                                      6                       6      6   
Ixelles                                   44                      44     44   
Jette                                     50                      50     50   
Koekelberg                                20                      20     20   
Laeken                                    18                      18     18   
Molenbeek-Saint-Jean                      36                      36     36   
Neder-Over-Heembeek                       13                      13     13   
Saint-Gilles                              38                      38     38   
Saint-Josse-Ten-Noode                     55                      55     55   
Schaerbeek                                36                      36     36   
Uccle                                     33                      33     33   
Watermael-Boitsfort                       37                      37     37   
Woluwe-Saint-Lambert                      44                      44     44   
Woluwe-Saint-Pierre                       20                      20     20   

                       Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                            
Anderlecht                         23               23              23  
Auderghem                          43               43              43  
Berchem-Sainte-Agathe              14               14              14  
Bruxelles                         100              100             100  
Etterbeek                          42               42              42  
Evere                              16               16              16  
Forest                             15               15              15  
Ganshoren                          30               30              30  
Haren                               6                6               6  
Ixelles                            44               44              44  
Jette                              50               50              50  
Koekelberg                         20               20              20  
Laeken                             18               18              18  
Molenbeek-Saint-Jean               36               36              36  
Neder-Over-Heembeek                13               13              13  
Saint-Gilles                       38               38              38  
Saint-Josse-Ten-Noode              55               55              55  
Schaerbeek                         36               36              36  
Uccle                              33               33              33  
Watermael-Boitsfort                37               37              37  
Woluwe-Saint-Lambert               44               44              44  
Woluwe-Saint-Pierre                20               20              20

In [33]:
print('There are {} uniques categories in Brussels.'.format(len(brussels_venues['Venue Category'].unique())))

There are 186 uniques categories in Brussels.


In the next part of the project I'll merge all the neighborhoods and segment them in clusters trying to find similar neghborhoods independently of the city they belong to.

# CapStone Project - week 5

After identifying and characterizing all the neighborhoods of Paris and Brussels, I'll now merge the data and segment the neighborhoods without knowing what city they belong to.

In [36]:
frames = [paris_venues, brussels_venues]
both_cities_df = pd.concat(frames)
both_cities_df.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0      Paris 1              48.860074                2.341092   
1      Paris 1              48.860074                2.341092   
2      Paris 1              48.860074                2.341092   
3      Paris 1              48.860074                2.341092   
4      Paris 1              48.860074                2.341092   

                                               Venue  Venue Latitude  \
0                              Cour Carrée du Louvre       48.860360   
1                                    Place du Louvre       48.859841   
2  Église Saint-Germain-l'Auxerrois (Église Saint...       48.859520   
3                                      Coffee Crêpes       48.858841   
4                                  Boutique yam'Tcha       48.861710   

   Venue Longitude      Venue Category  
0         2.338543    Pedestrian Plaza  
1         2.340822               Plaza  
2         2.341306              Church  
3         2.340802         Coffee Shop  
4         2.342380  Chinese Restaurant

In [37]:
# one hot encoding
both_cities_onehot = pd.get_dummies(both_cities_df[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
both_cities_onehot['Neighborhood'] = both_cities_df['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [both_cities_onehot.columns[-1]] + list(both_cities_onehot.columns[:-1])
both_cities_onehot = both_cities_onehot[fixed_columns]

both_cities_onehot.head()

Neighborhood  Afghan Restaurant  African Restaurant  American Restaurant  \
0      Paris 1                  0                   0                    0   
1      Paris 1                  0                   0                    0   
2      Paris 1                  0                   0                    0   
3      Paris 1                  0                   0                    0   
4      Paris 1                  0                   0                    0   

   Antique Shop  Arepa Restaurant  Argentinian Restaurant  Art Gallery  \
0             0                 0                       0            0   
1             0                 0                       0            0   
2             0                 0                       0            0   
3             0                 0                       0            0   
4             0                 0                       0            0   

   Art Museum  Arts & Crafts Store  ...  Vegetarian / Vegan Restaurant  \
0           0                    0  ...                              0   
1           0                    0  ...                              0   
2           0                    0  ...                              0   
3           0                    0  ...                              0   
4           0                    0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Volleyball Court  \
0                 0            0                      0                 0   
1                 0            0                      0                 0   
2                 0            0                      0                 0   
3                 0            0                      0                 0   
4                 0            0                      0                 0   

   Wine Bar  Wine Shop  Winery  Women's Store  Yoga Studio  
0         0          0       0              0            0  
1         0          0       0              0            0  
2         0          0       0              0            0  
3         0          0       0              0            0  
4         0          0       0              0            0  

[5 rows x 261 columns]

In [38]:
both_cities_onehot.shape

(2090, 261)

I'll now group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [39]:
both_cities_grouped = both_cities_onehot.groupby('Neighborhood').mean().reset_index()
both_cities_grouped

Neighborhood  Afghan Restaurant  African Restaurant  \
0              Anderlecht           0.000000            0.000000   
1               Auderghem           0.000000            0.000000   
2   Berchem-Sainte-Agathe           0.000000            0.000000   
3               Bruxelles           0.000000            0.000000   
4               Etterbeek           0.000000            0.000000   
5                   Evere           0.000000            0.000000   
6                  Forest           0.000000            0.000000   
7               Ganshoren           0.000000            0.000000   
8                   Haren           0.000000            0.000000   
9                 Ixelles           0.000000            0.022727   
10                  Jette           0.000000            0.000000   
11             Koekelberg           0.000000            0.000000   
12                 Laeken           0.000000            0.000000   
13   Molenbeek-Saint-Jean           0.000000            0.000000   
14    Neder-Over-Heembeek           0.000000            0.000000   
15                Paris 1           0.000000            0.000000   
16               Paris 10           0.000000            0.017544   
17               Paris 11           0.000000            0.011905   
18               Paris 12           0.000000            0.000000   
19               Paris 13           0.000000            0.000000   
20               Paris 14           0.000000            0.000000   
21               Paris 15           0.000000            0.000000   
22               Paris 16           0.000000            0.000000   
23               Paris 17           0.010101            0.000000   
24               Paris 18           0.000000            0.000000   
25               Paris 19           0.000000            0.000000   
26                Paris 2           0.000000            0.000000   
27               Paris 20           0.000000            0.000000   
28                Paris 3           0.000000            0.000000   
29                Paris 4           0.000000            0.010000   
30                Paris 5           0.000000            0.000000   
31                Paris 6           0.000000            0.000000   
32                Paris 7           0.000000            0.000000   
33                Paris 8           0.000000            0.000000   
34                Paris 9           0.000000            0.000000   
35           Saint-Gilles           0.000000            0.000000   
36  Saint-Josse-Ten-Noode           0.000000            0.018182   
37             Schaerbeek           0.000000            0.000000   
38                  Uccle           0.000000            0.000000   
39    Watermael-Boitsfort           0.000000            0.000000   
40   Woluwe-Saint-Lambert           0.000000            0.000000   
41    Woluwe-Saint-Pierre           0.000000            0.000000   

    American Restaurant  Antique Shop  Arepa Restaurant  \
0              0.000000      0.000000          0.000000   
1              0.000000      0.000000          0.000000   
2              0.000000      0.000000          0.000000   
3              0.010000      0.000000          0.000000   
4              0.000000      0.000000          0.000000   
5              0.000000      0.000000          0.000000   
6              0.000000      0.000000          0.000000   
7              0.000000      0.000000          0.000000   
8              0.000000      0.000000          0.000000   
9              0.000000      0.000000          0.000000   
10             0.000000      0.000000          0.000000   
11             0.000000      0.000000          0.000000   
12             0.000000      0.000000          0.000000   
13             0.000000      0.000000          0.000000   
14             0.000000      0.000000          0.000000   
15             0.000000      0.000000          0.000000   
16             0.000000      0.000000          0.000000   
17             0.000000   

Checking the top 5 venues of each neighborhood

In [40]:
num_top_venues = 5

for hood in both_cities_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = both_cities_grouped[both_cities_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Anderlecht----
                           venue  freq
0                     Food Court  0.09
1  Vegetarian / Vegan Restaurant  0.09
2             Italian Restaurant  0.09
3                            Pub  0.04
4              French Restaurant  0.04


----Auderghem----
                  venue  freq
0  Fast Food Restaurant  0.07
1       Thai Restaurant  0.05
2                   Bar  0.05
3           Pizza Place  0.05
4      Sushi Restaurant  0.05


----Berchem-Sainte-Agathe----
              venue  freq
0       Snack Place  0.14
1  Greek Restaurant  0.14
2       Sports Club  0.07
3               Gym  0.07
4               Bar  0.07


----Bruxelles----
            venue  freq
0        Beer Bar  0.09
1             Bar  0.06
2  Chocolate Shop  0.05
3     Coffee Shop  0.04
4         Gay Bar  0.04


----Etterbeek----
                venue  freq
0  Italian Restaurant  0.12
1         Pizza Place  0.07
2                 Bar  0.07
3    Greek Restaurant  0.05
4         Snack Place  0.05


----E

### Let's put that into a pandas dataframe
First, let's write a function to sort the venues in descending order.

In [41]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [42]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = both_cities_grouped['Neighborhood']

for ind in np.arange(both_cities_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(both_cities_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue          2nd Most Common Venue  \
0             Anderlecht    Italian Restaurant  Vegetarian / Vegan Restaurant   
1              Auderghem  Fast Food Restaurant                Thai Restaurant   
2  Berchem-Sainte-Agathe      Greek Restaurant                    Snack Place   
3              Bruxelles              Beer Bar                            Bar   
4              Etterbeek    Italian Restaurant                    Pizza Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Food Court     Convenience Store         Metro Station   
1    Italian Restaurant     French Restaurant           Snack Place   
2          Burger Joint            Restaurant                   Gym   
3        Chocolate Shop           Coffee Shop               Gay Bar   
4                   Bar                Bakery       Thai Restaurant   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Belgian Restaurant        Farmers Market           Supermarket   
1  Middle Eastern Restaurant           Pizza Place                   Bar   
2                       Park                   Bar           Supermarket   
3                      Plaza        Sandwich Place        Clothing Store   
4           Greek Restaurant           Snack Place                 Plaza   

  9th Most Common Venue 10th Most Common Venue  
0        Sandwich Place              Brasserie  
1      Sushi Restaurant       Insurance Office  
2              Pharmacy            Sports Club  
3             Bookstore                  Hotel  
4     French Restaurant           Concert Hall

In [48]:
neighborhoods_venues_sorted.shape

(42, 12)

In [43]:
both_cities_grouped.head()

Neighborhood  Afghan Restaurant  African Restaurant  \
0             Anderlecht                0.0                 0.0   
1              Auderghem                0.0                 0.0   
2  Berchem-Sainte-Agathe                0.0                 0.0   
3              Bruxelles                0.0                 0.0   
4              Etterbeek                0.0                 0.0   

   American Restaurant  Antique Shop  Arepa Restaurant  \
0                 0.00           0.0               0.0   
1                 0.00           0.0               0.0   
2                 0.00           0.0               0.0   
3                 0.01           0.0               0.0   
4                 0.00           0.0               0.0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  ...  \
0                     0.0          0.0         0.0                  0.0  ...   
1                     0.0          0.0         0.0                  0.0  ...   
2                     0.0          0.0         0.0                  0.0  ...   
3                     0.0          0.0         0.0                  0.0  ...   
4                     0.0          0.0         0.0                  0.0  ...   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                       0.086957               0.0          0.0   
1                       0.000000               0.0          0.0   
2                       0.000000               0.0          0.0   
3                       0.000000               0.0          0.0   
4                       0.000000               0.0          0.0   

   Vietnamese Restaurant  Volleyball Court  Wine Bar  Wine Shop  Winery  \
0                    0.0               0.0      0.00        0.0     0.0   
1                    0.0               0.0      0.00        0.0     0.0   
2                    0.0               0.0      0.00        0.0     0.0   
3                    0.0               0.0      0.01        0.0     0.0   
4                    0.0               0.0      0.00        0.0     0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 261 columns]

In [44]:
#both_cities_grouped_clustering = both_cities_grouped.drop('Neighborhood',1)
#both_cities_grouped_clustering.head()

### Cluster neighborhooods
Run k-means to cluster the neighborhoods in 8 clusters

In [45]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [61]:
# set number of clusters
kclusters = 8

both_cities_grouped_clustering = both_cities_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(both_cities_grouped_clustering)

# check cluster labels generated for each row in the dataframe
#kmeans.labels_[0:10] 
kmeans.labels_ 

array([0, 4, 4, 1, 5, 7, 4, 2, 6, 1, 2, 4, 4, 1, 2, 1, 1, 1, 1, 1, 3, 3,
       3, 1, 1, 3, 1, 1, 1, 1, 3, 1, 3, 3, 1, 1, 5, 2, 2, 5, 2, 4],
      dtype=int32)

In [62]:
neighborhoods_venues_sorted.drop(['Cluster Labels'], axis=1, inplace = True)

In [63]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

both_cities_merged = both_cities_neigh_df

# merge both_cities_grouped with both_cities_df to add latitude/longitude for each neighborhood
both_cities_merged = both_cities_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

both_cities_merged.head() # check the last columns!

Postal Code Neighborhood   Latitude  Longitude  Cluster Labels  \
0        75001      Paris 1  48.860074   2.341092               1   
1        75002      Paris 2  48.866729   2.340470               1   
2        75003      Paris 3  48.863873   2.361786               1   
3        75004      Paris 4  48.856053   2.356322               1   
4        75005      Paris 5  48.846198   2.344591               3   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0     French Restaurant                  Café                 Plaza   
1     French Restaurant   Japanese Restaurant              Wine Bar   
2     French Restaurant                 Hotel           Art Gallery   
3     French Restaurant        Clothing Store          Gourmet Shop   
4     French Restaurant                 Hotel                   Bar   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                   Bar                 Hotel              Wine Bar   
1                Bistro        Clothing Store    Italian Restaurant   
2   Japanese Restaurant          Cocktail Bar    Italian Restaurant   
3        Ice Cream Shop              Wine Bar        Cosmetics Shop   
4    Italian Restaurant                   Pub                  Café   

  7th Most Common Venue  8th Most Common Venue 9th Most Common Venue  \
0           Pizza Place                Theater            Restaurant   
1                  Café               Boutique                 Hotel   
2              Wine Bar  Vietnamese Restaurant                Bistro   
3                 Hotel                Gay Bar                  Café   
4   Indie Movie Theater         Ice Cream Shop                 Plaza   

  10th Most Common Venue  
0            Cheese Shop  
1                 Bakery  
2           Burger Joint  
3     Italian Restaurant  
4     Chinese Restaurant

Finally let's visualize the data around both Paris and Brussels

In [56]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [64]:
# create map
map_clusters_p = folium.Map(location=[latitude_p, longitude_p], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(both_cities_merged['Latitude'], both_cities_merged['Longitude'], both_cities_merged['Neighborhood'], both_cities_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters_p)
       
map_clusters_p

In [65]:
# create map
map_clusters_b = folium.Map(location=[latitude_b, longitude_b], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(both_cities_merged['Latitude'], both_cities_merged['Longitude'], both_cities_merged['Neighborhood'], both_cities_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters_b)
       
map_clusters_b

# Examine clusters

### Cluster 1

In [68]:
both_cities_merged.loc[both_cities_merged['Cluster Labels'] == 0, both_cities_merged.columns[[1] + list(range(5, both_cities_merged.shape[1]))]]

Neighborhood 1st Most Common Venue          2nd Most Common Venue  \
6   Anderlecht    Italian Restaurant  Vegetarian / Vegan Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6            Food Court     Convenience Store         Metro Station   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6    Belgian Restaurant        Farmers Market           Supermarket   

  9th Most Common Venue 10th Most Common Venue  
6        Sandwich Place              Brasserie

### Cluster 2

In [69]:
both_cities_merged.loc[both_cities_merged['Cluster Labels'] == 1, both_cities_merged.columns[[1] + list(range(5, both_cities_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0                Paris 1     French Restaurant                  Café   
1                Paris 2     French Restaurant   Japanese Restaurant   
2                Paris 3     French Restaurant                 Hotel   
3                Paris 4     French Restaurant        Clothing Store   
5                Paris 6     French Restaurant    Italian Restaurant   
8                Paris 9     French Restaurant                 Hotel   
9               Paris 10           Coffee Shop           Pizza Place   
10              Paris 11     French Restaurant                   Bar   
11              Paris 12                Bistro                 Hotel   
12              Paris 13                 Hotel     French Restaurant   
16              Paris 17                   Bar     French Restaurant   
17              Paris 18     French Restaurant                   Bar   
19              Paris 20                   Bar                Bakery   
0              Bruxelles              Beer Bar                   Bar   
4                Ixelles    Italian Restaurant                   Bar   
5           Saint-Gilles                   Bar             Brasserie   
7   Molenbeek-Saint-Jean                   Bar           Art Gallery   

   3rd Most Common Venue  4th Most Common Venue  \
0                  Plaza                    Bar   
1               Wine Bar                 Bistro   
2            Art Gallery    Japanese Restaurant   
3           Gourmet Shop         Ice Cream Shop   
5                   Café         Sandwich Place   
8    Japanese Restaurant   Gym / Fitness Center   
9           Cocktail Bar      French Restaurant   
10                  Café                 Bistro   
11     French Restaurant     Chinese Restaurant   
12       Thai Restaurant  Vietnamese Restaurant   
16              Wine Bar     Italian Restaurant   
17           Pizza Place                  Hotel   
19     French Restaurant    Japanese Restaurant   
0         Chocolate Shop            Coffee Shop   
4               Tea Room            Coffee Shop   
5     Italian Restaurant                  Plaza   
7            Yoga Studio            Coffee Shop   

            5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                           Hotel              Wine Bar           Pizza Place   
1                  Clothing Store    Italian Restaurant                  Café   
2                    Cocktail Bar    Italian Restaurant              Wine Bar   
3                        Wine Bar        Cosmetics Shop                 Hotel   
5                           Plaza                Bakery           Tailor Shop   
8                     Coffee Shop              Wine Bar                   Bar   
9              Seafood Restaurant        Breakfast Spot                Bakery   
10                   Cocktail Bar                Bakery    Italian Restaurant   
11                    Supermarket      Tapas Restaurant     Health Food Store   
12                            Bar     Indian Restaurant    Italian Restaurant   
16                          Hotel  Gym / Fitness Center            Restaurant   
17                       Wine Bar             Gastropub           Supermarket   
19                           Café                 Plaza                Bistro   
0                         Gay Bar                 Plaza        Sandwich Place   
4   Vegetarian / Vegan Restaurant            Beer Store             Pool Hall   
5                     Pizza Place              Beer Bar     French Restaurant   
7                     Tailor Shop    Belgian Restaurant           Supermarket   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Theater            Restaurant            Cheese Shop  
1                Boutique                 Hotel                 Bakery  
2   Vietnamese Restaurant                Bistro           Burger Joint  
3                 Gay Bar                  Ca

### Cluster 3

In [71]:
both_cities_merged.loc[both_cities_merged['Cluster Labels'] == 2, both_cities_merged.columns[[1] + list(range(5, both_cities_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
2             Schaerbeek          Tram Station                Bakery   
10             Ganshoren                Bakery                  Park   
11                 Jette                Bakery             Pet Store   
12   Neder-Over-Heembeek      Greek Restaurant                  Park   
18                 Uccle                Bakery                 Plaza   
20  Woluwe-Saint-Lambert        Clothing Store        Chocolate Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2                  Plaza           Supermarket           Snack Place   
10     Convenience Store              Pharmacy        Shop & Service   
11        Cosmetics Shop  Gym / Fitness Center    Italian Restaurant   
12              Bus Stop                Bakery             Bookstore   
18        Sandwich Place        Cosmetics Shop           Supermarket   
20             Bookstore          Gourmet Shop         Jewelry Store   

   6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
2              Gastropub              Bus Stop  Middle Eastern Restaurant   
10    Belgian Restaurant           Supermarket             Sandwich Place   
11           Supermarket                   Bar                Pizza Place   
12           Flower Shop           Supermarket          Convenience Store   
18         Women's Store    Chinese Restaurant         Italian Restaurant   
20           Salad Place      Sushi Restaurant                Supermarket   

   9th Most Common Venue 10th Most Common Venue  
2     Spanish Restaurant    Sporting Goods Shop  
10             Bookstore               Boutique  
11                 Plaza               Creperie  
12              Friterie       Doner Restaurant  
18      Swiss Restaurant      Food & Drink Shop  
20             Brasserie    Sporting Goods Shop

### Cluster 4

In [72]:
both_cities_merged.loc[both_cities_merged['Cluster Labels'] == 3, both_cities_merged.columns[[1] + list(range(5, both_cities_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
4       Paris 5     French Restaurant                 Hotel   
6       Paris 7     French Restaurant                 Hotel   
7       Paris 8                 Hotel     French Restaurant   
13     Paris 14     French Restaurant    Italian Restaurant   
14     Paris 15     French Restaurant    Italian Restaurant   
15     Paris 16     French Restaurant                Bakery   
18     Paris 19     French Restaurant           Supermarket   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4                    Bar    Italian Restaurant                   Pub   
6                   Café    Italian Restaurant                Bakery   
7     Italian Restaurant                  Café                Bistro   
13                   Bar                Bakery      Sushi Restaurant   
14                 Hotel           Coffee Shop                   Bar   
15                  Café            Restaurant                 Hotel   
18    Italian Restaurant                  Pool                   Bar   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4                   Café   Indie Movie Theater        Ice Cream Shop   
6            Coffee Shop            Art Museum            Restaurant   
7                    Pub      Sushi Restaurant                   Bar   
13                 Hotel                Bistro             Bookstore   
14   Japanese Restaurant                  Park   Lebanese Restaurant   
15         Train Station                Garden                   Spa   
18                Bakery          Dessert Shop          Concert Hall   

   9th Most Common Venue 10th Most Common Venue  
4                  Plaza     Chinese Restaurant  
6                 Bistro                   Park  
7                 Bakery             Restaurant  
13            Restaurant            Fish Market  
14             Brasserie   Gym / Fitness Center  
15           Pizza Place     Chinese Restaurant  
18        Scenic Lookout            Beer Garden

### Cluster 5

In [73]:
both_cities_merged.loc[both_cities_merged['Cluster Labels'] == 4, both_cities_merged.columns[[1] + list(range(5, both_cities_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
1                  Laeken    Italian Restaurant          Burger Joint   
8              Koekelberg                   Gym     Convenience Store   
9   Berchem-Sainte-Agathe      Greek Restaurant           Snack Place   
15    Woluwe-Saint-Pierre  Fast Food Restaurant           Supermarket   
16              Auderghem  Fast Food Restaurant       Thai Restaurant   
19                 Forest           Supermarket                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1     Athletics & Sports        Sandwich Place         Metro Station   
8         Sandwich Place                Hostel           Bus Station   
9           Burger Joint            Restaurant                   Gym   
15    Italian Restaurant                 Plaza    Athletics & Sports   
16    Italian Restaurant     French Restaurant           Snack Place   
19     Health Food Store                   Bar      Halal Restaurant   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1                 Supermarket            Restaurant           Sports Club   
8                Soccer Field           Snack Place           Supermarket   
9                        Park                   Bar           Supermarket   
15          French Restaurant          Gourmet Shop      Greek Restaurant   
16  Middle Eastern Restaurant           Pizza Place                   Bar   
19       Gym / Fitness Center         Train Station          Tram Station   

   9th Most Common Venue 10th Most Common Venue  
1             Hookah Bar           Tram Station  
8     Falafel Restaurant               Tea Room  
9               Pharmacy            Sports Club  
15                Bistro             Restaurant  
16      Sushi Restaurant       Insurance Office  
19                Bistro                 Market

### Cluster 6

In [74]:
both_cities_merged.loc[both_cities_merged['Cluster Labels'] == 5, both_cities_merged.columns[[1] + list(range(5, both_cities_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
3               Etterbeek    Italian Restaurant           Pizza Place   
17    Watermael-Boitsfort            Restaurant    Italian Restaurant   
21  Saint-Josse-Ten-Noode    Italian Restaurant        Sandwich Place   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
3                    Bar                Bakery       Thai Restaurant   
17     French Restaurant        Ice Cream Shop        Sandwich Place   
21            Restaurant           Pizza Place                 Plaza   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3       Greek Restaurant           Snack Place                 Plaza   
17    Chinese Restaurant             Gastropub         Deli / Bodega   
21    Belgian Restaurant       Thai Restaurant          Concert Hall   

   9th Most Common Venue 10th Most Common Venue  
3      French Restaurant           Concert Hall  
17     Food & Drink Shop                   Park  
21        Ice Cream Shop                    Bar

### Cluster 7

In [75]:
both_cities_merged.loc[both_cities_merged['Cluster Labels'] == 6, both_cities_merged.columns[[1] + list(range(5, both_cities_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
13        Haren              Bus Stop           Snack Place   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
13           Comedy Club         Train Station    Athletics & Sports   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
13              Fountain           Flower Shop     Food & Drink Shop   

   9th Most Common Venue 10th Most Common Venue  
13            Food Court           Food Service

### Cluster 8

In [76]:
both_cities_merged.loc[both_cities_merged['Cluster Labels'] == 7, both_cities_merged.columns[[1] + list(range(5, both_cities_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
14        Evere                Bakery                Market   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
14            Sports Bar                  Park          Tennis Court   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14        Sandwich Place           Supermarket             Brasserie   

   9th Most Common Venue 10th Most Common Venue  
14          Hockey Field             Restaurant